src/graph/sampler/metapath.cc

```diff
@@ -0,0 +1,102 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file graph/sampler/metapath.cc
+ * \brief Metapath sampling
+ */
+
+#include <dgl/array.h>
+#include <dgl/random.h>
+#include <dgl/packed_func_ext.h>
+#include "../../c_api_common.h"
+#include "randomwalk.h"
+
+using namespace dgl::runtime;
+using namespace dgl::aten;
+
+namespace dgl {
+
+namespace sampling {
+
+namespace {
+
+/*!
+ * \brief Random walk based on indefinite cycle of given metapaths.
+ *
+ * \param hg The heterograph
+ * \param etypes The metapath as an array of edge type IDs
+ * \param seeds The array of starting vertices for random walks
+ * \param num_traces Number of traces to generate for each starting vertex
+ * \param max_cycles Maximum number of metapath traversals for each starting vertex
+ * \note The metapath should have the same starting and ending node type.
+ */
+RandomWalkTracesPtr MetapathCycleRandomWalk(
+    const HeteroGraphPtr hg,
+    const IdArray etypes,
+    const IdArray seeds,
+    int num_traces,
+    int max_cycles) {
+  const auto metagraph = hg->meta_graph();
+  uint64_t num_etypes = etypes->shape[0];
+  uint64_t num_seeds = seeds->shape[0];
+  const dgl_type_t *etype_data = static_cast<dgl_type_t *>(etypes->data);
+  const dgl_id_t *seed_data = static_cast<dgl_id_t *>(seeds->data);
+
+  std::vector<dgl_id_t> vertices;
+  std::vector<size_t> trace_lengths, trace_counts;
+
+  for (uint64_t seed_id = 0; seed_id < num_seeds; ++seed_id) {
```

Comment:
Consider using omp to embarrassingly parallelize this loop? If so, you might need to trigger `SuccVec` first to materialize the lazy component before entering the critical area. You might also need to split the `vertices` vector into ones for each seed.

Reply:
Sounds not very trivial to implement...  Can I postpone OpenMP parallelization to 0.5?

Reply:
sure. added a TODO note there